lightning-transaction-sync/src/esplora.rs

```diff
@@ -0,0 +1,373 @@
+use crate::TxSyncError;
+
+use lightning::util::logger::Logger;
+use lightning::{log_error, log_given_level, log_info, log_internal, log_debug, log_trace};
+use lightning::chain::WatchedOutput;
+use lightning::chain::{Confirm, Filter};
+
+use bitcoin::{BlockHash, BlockHeader, Script, Transaction, Txid};
+
+use esplora_client::Builder;
+#[cfg(feature = "async-interface")]
+use esplora_client::r#async::AsyncClient;
+#[cfg(not(feature = "async-interface"))]
+use esplora_client::blocking::BlockingClient;
+
+use std::sync::Mutex;
+use std::collections::HashSet;
+use core::ops::Deref;
+
+/// Synchronizes LDK with a given [`Esplora`] server.
+///
+/// Needs to be registerd with a [`ChainMonitor`] via the [`Filter`] interface to be informed of
+/// transactions and outputs to montor.
+///
+/// [`Esplora`]: https://github.com/Blockstream/electrs
+/// [`ChainMonitor`]: lightning::chain::chainmonitor::ChainMonitor
+/// [`Filter`]: lightning::chain::Filter
+pub struct EsploraSyncClient<L: Deref>
+where
+	L::Target: Logger,
+{
+	// Transactions that were registered via the `Filter` interface and have to be processed.
+	queued_transactions: Mutex<HashSet<Txid>>,
+	// Transactions that were previously processed, but must not be forgotten yet.
+	watched_transactions: Mutex<HashSet<Txid>>,
+	// Outputs that were registered via the `Filter` interface and have to be processed.
+	queued_outputs: Mutex<HashSet<WatchedOutput>>,
+	// Outputs that were previously processed, but must not be forgotten yet.
+	watched_outputs: Mutex<HashSet<WatchedOutput>>,
+	// The tip hash observed during our last sync.
+	#[cfg(not(feature = "async-interface"))]
+	last_sync_hash: std::sync::Mutex<Option<BlockHash>>,
+	#[cfg(feature = "async-interface")]
+	last_sync_hash: futures::lock::Mutex<Option<BlockHash>>,
+	#[cfg(not(feature = "async-interface"))]
+	client: BlockingClient,
+	#[cfg(feature = "async-interface")]
+	client: AsyncClient,
+	logger: L,
+}
+
+impl<L: Deref> EsploraSyncClient<L>
+where
+	L::Target: Logger,
+{
+	/// Synchronizes the given confirmables via the [`Confirm`] interface. This method should be
+	/// called regularly to keep LDK up-to-date with current chain data.
+	///
+	/// [`Confirm`]: lightning::chain::Confirm
+	#[maybe_async]
```

Comment:
Actually, the docs of `bdk_macros` are bit misleading here `maybe_await`/`maybe_async` function on the basis of 
```rust
#[cfg(any(target_arch = "wasm32", feature = "async-interface"))]
```

So the methods will be made `async`/awaited when *either*  the arch is `wasm32` *or* the `async-interface` feature is set. Not entirely sure about the WASM part, but the parts where BDK uses these macros seem to work fine for Mutiny at least? If we want to change that/something we can def. fork and create our own `ldk_macros` crate, just thought this might be a bit redundant for the time being.

Reply:
> Ah, we should not be setting it in wasm32, that's...a really strange API quirk.

Hm, so far my understanding was that WASM *needs* to use the async API as there are no threads. So why wouldn't we want to automatically switch it over on `wasm32`?

Reply:
> Thats not strictly true, in theory you can run things in web workers to get multiple OS threads. More generally, this crate doesn't really make sense in wasm either way - it has to make an HTTP request, which we can't assume works from within wasm without some additional glue. I believe BDK is assuming that users are using wasm_bindgen, with a bunch of glue code being hidden there to call JS and have that make the HTTP call.

Actually, the `wasm_bindgen` part is  not happening in BDK, but downstream in `reqwest`, which is used by the async Esplora client. As we're currently dependent on this anyways, I'm still not convinced we can't go the same route, since it seems to work fine for BDK?

> In general, IMO it would be nice to do what the user says - if they ask for async, build async, if they don't, don't. In the future we should probably have a more flexible approach to HTTP, which would let us build this crate with our own wasm bindings without the complexity and overhead of wasm_bindegn.

Right, but that would require rewriting `rust-esplora-client`, or at least providing an entirely new third variant (besides `BlockingClient` and `AsyncClient`) that would allow to handle any request based on a custom HTTP client implementation. Not sure we need/want to do this right away/soon?

Reply:
Alright, just discussed the matter with the BDK devs, and they're open to changing the macros crate, so I opened a PR (https://github.com/bitcoindevkit/.github/issues/39), which will be discussed in BDK Lib call next week.